ducc/src/object.rs

```diff
@@ -211,16 +224,80 @@ impl<'ducc> Object<'ducc> {
     }
 }
 
-#[derive(Default)]
-pub struct PropertyDescriptor<'ducc, V>
-where
-    V: ToValue<'ducc> {
-    pub enumerable: bool,
-    pub configurable: bool,
-    pub writable: bool,
-    pub value: Option<V>,
-    pub get: Option<Function<'ducc>>,
-    pub set: Option<Function<'ducc>>,
+enum PropertySource<'ducc> {
+    Undefined,
+    Value(Value<'ducc>),
+    GetSet(Function<'ducc>, Function<'ducc>),
+    Get(Function<'ducc>),
+    Set(Function<'ducc>),
+}
+
+pub struct PropertyDescriptor<'ducc> {
+    enumerable: Option<bool>,
+    configurable: Option<bool>,
+    writable: Option<bool>,
+    source: PropertySource<'ducc>
+}
+impl <'ducc> PropertyDescriptor<'ducc> {
+    pub fn new() -> PropertyDescriptor<'ducc> {
+        PropertyDescriptor {
+            enumerable: None,
+            configurable: None,
+            writable: None,
+            source: PropertySource::Undefined
+        }
+    }
+
+    /// Whether this property shows up during enumeration of the
+    /// properties on the corresponding object.
+    /// 
+    /// Defaults to `false`
+    pub fn enumerable<'a>(&'a mut self, b: bool) -> &'a mut PropertyDescriptor<'ducc> {
+        self.enumerable = Some(b);
+        self
+    }
+
+    /// Whether the type of this property descriptor may be changed and
+    /// the property may be deleted from the corresponding object.
+    /// 
+    /// Defaults to `false`
+    pub fn configurable<'a>(&'a mut self, b: bool) -> &'a mut PropertyDescriptor<'ducc> {
+        self.configurable = Some(b);
+        self
+    }
+
+    /// Whether the value associated with the property may be changed with
+    /// an assignment operator. Must not be set when using getters or setters.
```

Comment:
Correct me if my understanding is wrong. The API you've made, with all of the `Option<bool>` fields, allows not only for creating properties but also updating. Therefore it's not straightforward to disallow setting `writable` erroneously, because perhaps you want to be able to update a property to `writable = true` without also specifying `source = Value(...)` in the update.

This error (actually it's a panic, right?) can be resolved by:

* Only allowing `writable()` when also specifying a `value()`.
* Calling `duk_get_prop_desc` when both `writable` is not `None` and `source` is `Undefined` in order to verify that the property is indeed backed by a `value`.

Or you can just leave it as a panic. In that case I would just change the documentation to

```
/// # Panics
///
/// Panics if `writable` is set when defining or updating a property that uses getters or setters.
```

and I would add a `#[should_panic]` test for the condition.

Reply:
The intention was to try to stick close to what Duktape API allows, since the flags passed to `duk_def_prop` allow you to specify both whether you have passed the value for e.g. `configurable` and the value itself. Therefore my thought process was that `Option<bool>` allows specifying both whether the flag was provided and what the flag was. I didn't think about the updating use case at all when implementing this tbh.

While you can create an invalid `PropertyDescriptor` that will cause duk_def_prop to return an error, the error will be [handled in Duktape itself](https://github.com/svaarala/duktape/blob/c91bcc9b1c05f65b2f6f7454faa43388ed92fb1d/src-input/duk_api_object.c#L547). That's why I didn't add a separate panic/error handling. I don't really have strong opinion whether it's better to handle the configuration error in Ducc configuration builder or Duktape API though.

Reply:
Oh, for some reason I was thinking that it was a *fatal* Duktape error (which would cause a panic). But it's just raised as a JavaScript exception? No worries then! I think this is fine as-is.